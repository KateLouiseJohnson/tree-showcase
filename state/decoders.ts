import { Decoder } from 'elm-decoders';
import { Tree } from '../types'

// Decoder is used here to validate the data and then serialize
const treeDecoder: Decoder<Tree> = Decoder.object({
  name: Decoder.string,
  species_name: Decoder.string,
  image: Decoder.string,
}).map(({name, species_name, image}) => ({
  name,
  species: species_name,
  imageUrl: image,
}))

export const treesDecoder = Decoder.object({
  trees: Decoder.array(treeDecoder)
})
