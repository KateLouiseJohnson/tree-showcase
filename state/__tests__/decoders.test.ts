import {describe, expect, it} from '@jest/globals'
import { treesDecoder } from '../decoders'
import { apiResponsePayload } from '../../fixtures/apiResponsePayload'

describe('treesDecoder', () => {
  it('should decode response from api', () => {
    const result = treesDecoder.run(apiResponsePayload)
    expect(result).toMatchSnapshot()
  })
  
  it('should fail for invalid data', () => {
    const result = treesDecoder.run({trees: [{}]})
    expect(result).toMatchSnapshot()
  })
})
