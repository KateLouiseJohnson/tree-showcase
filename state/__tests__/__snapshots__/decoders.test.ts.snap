// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treesDecoder should decode response from api 1`] = `
Object {
  "type": "OK",
  "value": Object {
    "trees": Array [
      Object {
        "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/3/36/Baobab_Adansonia_digitata.jpg",
        "name": "Baobab",
        "species": "Adansonia",
      },
      Object {
        "imageUrl": "https://upload.wikimedia.org/wikipedia/en/1/16/Red_mangrove-everglades_natl_park.jpg",
        "name": "Red Mangrove",
        "species": "Rhizophora mangle",
      },
      Object {
        "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/2/2c/Carpinus_betulus_-_Hunsr%C3%BCck_001.jpg",
        "name": "Common Hornbeam",
        "species": "Carpinus betulus",
      },
      Object {
        "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/3/34/Quercus_cerris.JPG",
        "name": "Turkey Oak",
        "species": "Quercus cerris",
      },
      Object {
        "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/f/f0/Pinus_syluestriformis_%28Takenouchi%29T.Wang_ex_Cheng.JPG",
        "name": "Japanese red pine",
        "species": "Pinus densiflora",
      },
    ],
  },
}
`;

exports[`treesDecoder should fail for invalid data 1`] = `
Object {
  "error": Object {
    "trees": Array [
      Array [
        0,
        Object {
          "image": Object {
            "error": "Not a string",
            "value": undefined,
          },
          "name": Object {
            "error": "Not a string",
            "value": undefined,
          },
          "species_name": Object {
            "error": "Not a string",
            "value": undefined,
          },
        },
      ],
    ],
  },
  "type": "FAIL",
}
`;
